### ISSUE TITLE
Bug: @typescript-eslint/no-unused-vars false positive when using "using" declaration statement

### ISSUE THREAD

**Issue by jaenster** [typescript-eslint/typescript-eslint#9175]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.5&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXzDigxABkcwBrACgEoAueABRhwFssBnEAHgG94AbQDKATzYAjHBAB0UTqPQARLgAcc3ALp1GLdl14A3HFmAA%2BeAF8zAbgBQd%2BYrBI0mXPmTdyVOvD528PAA9MHwAGwAjADM0YwA5BAUlPHwXPCgkLAgwPASyBhpGJzwhlAQyAjpqCCGIDDwRMAy8ACs9JGt8AC0bR0AnIHwUADuUFiFXlioAObwSVTwALzDYxMERCQ%2BNLT2Q6Hw1JzsCGA4oHkLlPAAFvUgtA5BBwCCJRg36TMgxfAYMFBUJxVFgIDk-jh4Mc2D9PrN5lhKAgPlx9gAqIbzH7zZI2eCY-6ifyYoI4xYrUbjQqEEDEMjJOiYyxIablCBEgKk0mU9ZXAD8MhE4iksicyjUGhA2l2TKGaOC%2BzCZRgWCgEjBZOuwBwIBK30m3GAdksQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare function createLock(): Promise<{ [Symbol.asyncDispose](): Promise<void> }>;

async function useLock() {
  // 6133: 'lock' is declared but its value is never read. 5:15 - 5:19
  await using lock = await createLock();

  // (some code block here)

  // As this gets transpiled to something like this
  /*
   let lock; 
   try {
      lock = await createLock()
   } finally {
      await lock?.[Symbol.asyncDispose]();
   }
  */
  // variable lock does get used
}
``
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    '@typescript-eslint/no-unused-vars': ["error"],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expected that eslint understands that using a "using" declaration implies that that variable is used, always on the end of th block

### Actual Result

6133: 'x' is declared but its value is never read.

**Comment by bradzacher** [comment]
I don't necessarily agree here. This is implicitly used, yes, but it's not explicitly used. There is a side-effect to a keyword, but the variable is still unused.

I'm personally of the opinion that the rule should report here and that you should explicitly annotate the variable as intentionally unused (eg via disable comment or via the ignore option naming convention).

As an example - in rust data types can have a `Drop` implementation which is called when the data is dropped from memory. It's common to use the drop implementation to free resources like file handles. So long as the value is retained in scope the handle is alive. So it's common for utilities to return the value an document that the variable isn't meant to be used and is instead just there as a marker for when you're done with the handle. Clippy (rust's linter) doesn't encode any knowledge about this - and so the convention is to declare the variable with a leading underscore to denote it as intentionally unused (silencing the linter).
I see this as exactly the same case.

As a concrete example of why I think this should remain as is. Imagine your util instead returns a disposable DB connection.
If that variable is unused you'd want to delete it - no sense opening a DB connection and then disposing it later if you aren't going to use it. So if the rule ignores `using` then you'd never get a lint error telling you to clean the variable up.

So yeah. I think this should remain as a reported pattern and a configured naming convention should be used to mark the variable as intentionally unused.

I'd like to hear thoughts from others. I've yet to use the feature so I don't have any real experience or ideas here - but I don't see why it should be treated differently.  @typescript-eslint/triage-team

**Comment by jaenster** [comment]
> I don't necessarily agree here. This is implicitly used, yes, but it's not explicitly used. There is a side-effect to a keyword, but the variable is still unused.
> 
> I'm personally of the opinion that the rule should report here and that you should explicitly annotate the variable as intentionally unused (eg via disable comment or via the ignore option naming convention).
> 
> As an example - in rust data types can have a `Drop` implementation which is called when the data is dropped from memory. It's common to use the drop implementation to free resources like file handles. So long as the value is retained in scope the handle is alive. So it's common for utilities to return the value an document that the variable isn't meant to be used and is instead just there as a marker for when you're done with the handle. Clippy (rust's linter) doesn't encode any knowledge about this - and so the convention is to declare the variable with a leading underscore to denote it as intentionally unused (silencing the linter). I see this as exactly the same case.
> 
> As a concrete example of why I think this should remain as is. Imagine your util instead returns a disposable DB connection. If that variable is unused you'd want to delete it - no sense opening a DB connection and then disposing it later if you aren't going to use it. So if the rule ignores `using` then you'd never get a lint error telling you to clean the variable up.
> 
> So yeah. I think this should remain as a reported pattern and a configured naming convention should be used to mark the variable as intentionally unused.
> 
> I'd like to hear thoughts from others. I've yet to use the feature so I don't have any real experience or ideas here - but I don't see why it should be treated differently. @typescript-eslint/triage-team

typically people use it with something like say a file handler, so typically people will use the file right, because otherwise what is the point. Still there is an option to disable it for function arguments, this might be one of these settings that is required, because at the end of the day it is used right. Like with function arguments you can say you need to implement them because you implement some method signature.

Maybe that is the solution, having the option to disable it for with:
`'@typescript-eslint/no-unused-vars': ["error", { "using": "none" }]`

But i'm just thinking out loud here, i do get your point too. A unused variable implies that it is a waste of variable, a pointless statement. You either have to replace it with an expression statement or remove all together, but with this using example it is not a pointless statement, but eslint suggest it is

**Comment by bradzacher** [comment]
I never actually clicked into your playground.

That error isn't from eslint. It's from typescript.
Note how the section is entitled "TypeScript" and how you never actually turned on any eslint rules.

If you turn on our `no-unused-vars` lint rule then it reports on the variable all the same. [playground](https://typescript-eslint.io/play/#ts=5.4.5&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXzDigxABkcwBrACgEoAueABRhwFssBnEAHgG94AbQDKATzYAjHBAB0UTqPQARLgAcc3ALp1GLdl14A3HFmAA%2BeAF8zAbgBQd%2BYrBI0mXPmTdyVOvD528PAA9MHwAGwAjADM0YwA5BAUlPHwXPCgkLAgwPASyBhpGJzwhlAQyAjpqCCGIDDwRMAy8ACs9JGt8AC0bR0AnIHwUADuUFiFXlioAObwSVTwALzDYxMERCQ%2BNLT2Q6Hw1JzsCGA4oHkLlPAAFvUgtA5BBwCCJRg36TMgxfAYMFBUJxVFgIDk-jh4Mc2D9PrN5lhKAgPlx9gAqIbzH7zZI2eCY-6ifyYoI4xYrUbjQqEEDEMjJOiYyxIablCBEgKk0mU9ZXAD8MhE4iksicyjUGhA2l2TKGaOC%2BzCZRgWCgEjBZOuwBwIBK30m3GAdksQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

Which further suggests that our behaviour is correct - TS agrees with us!

**Comment by bradzacher** [comment]
> A unused variable implies that it is a waste of variable, a pointless statement. You either have to replace it with an expression statement or remove all together, but with this using example it is not a pointless statement, but eslint suggest it is

ESLint doesn't say the variable is pointless - it merely tells you the variable is never used:

```
'lock' is assigned a value but never used.
```

Replacing it with an expression statement is one valid resolution if you want to _remove_ the variable. Using the variable in some way is another option (if you forgot to use it). Or naming the variable according to the `ignore` configuration is another option.

There's a number of resolutions depending on what you intend.

As a related example - it's possible to configure the rule to report on all function args. But sometimes you're implementing an interface or a callback and expecting to purposely not use, say, the first arg and only use the second arg. In this case you similarly can't delete the first defined parameter as it's part of an external API constraint. So the only valid resolution is to name the intentionally unused parameter based on the `ignore` configuration.

**Comment by jaenster** [comment]
> I never actually clicked into your playground.
> 
> That error isn't from eslint. It's from typescript. Note how the section is entitled "TypeScript" and how you never actually turned on any eslint rules.
> 
> If you turn on our `no-unused-vars` lint rule then it reports on the variable all the same. [playground](https://typescript-eslint.io/play/#ts=5.4.5&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXzDigxABkcwBrACgEoAueABRhwFssBnEAHgG94AbQDKATzYAjHBAB0UTqPQARLgAcc3ALp1GLdl14A3HFmAA%2BeAF8zAbgBQd%2BYrBI0mXPmTdyVOvD528PAA9MHwAGwAjADM0YwA5BAUlPHwXPCgkLAgwPASyBhpGJzwhlAQyAjpqCCGIDDwRMAy8ACs9JGt8AC0bR0AnIHwUADuUFiFXlioAObwSVTwALzDYxMERCQ%2BNLT2Q6Hw1JzsCGA4oHkLlPAAFvUgtA5BBwCCJRg36TMgxfAYMFBUJxVFgIDk-jh4Mc2D9PrN5lhKAgPlx9gAqIbzH7zZI2eCY-6ifyYoI4xYrUbjQqEEDEMjJOiYyxIablCBEgKk0mU9ZXAD8MhE4iksicyjUGhA2l2TKGaOC%2BzCZRgWCgEjBZOuwBwIBK30m3GAdksQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
> 
> Which further suggests that our behaviour is correct - TS agrees with us!

Right! i totally missed that, sorry, i encountered the actual eslint error in my own code base that i didnt put up in the playground ofc, my bad :)

Thanks for giving a proper/working example. Still normally you can safely remove unused variables, but in this case it actively changes the behavior of your code by removing it. Much like the example of unused params, removing it changes the signature, so it actively changes behavior if you remove it.

While i agree TS agrees with the error i would say also typescript is wrong, specially typescript more so as eslint as that does tranpile it to something where it uses it.

I go ahead and transpiled this example code on the ts [playground](https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzDigxABkcwBrACgEoAueABRhwFssBnEAHgG94AbQDKATzYAjHBAB0UTqPQARLgAcc3ALp1GLdl14A3HFmAA+eAF8zAbgBQd+YrBI0mXPmTdyVOvD528PAA9MHwAGwAjADM0YwA5BAUlPHwXPCgkLAgwPASyBhpGJzwhlAQyAjpqCCGIDDwRMAy8ACs9JGt8AC0bR0AnIHwUADuUFiFXlioAObwSVTwALzDYxMERCQ+NLT2Q6Hw1JzsCGA4oHkLlPAAFvUgtA5BBwCCJRg36TMgxfAYMFBUJxVFgIDk-jh4Mc2D9PrN5lhKAgPlx9gAqIbzH7zZI2eCY-6ifyYoI4xYrUbjQqEEDEMjJOiYyxIablCBEgKk0mU9ZXAD8MhE4iksicyjUGhA2l2TKGaOC+zCZRgWCgEjBZOuwBwIBK30m3GAdksQA) and to my surprise it doesn't use the lock variable at all.

It might be a non issue indeed, still it is perfectly valid code and i dont think i run into the issue alone, it is a pretty common pattern to have a lock unlock itself when it goes out of scope, like in cpp with mutex's where it uses the strength of a deconstructor being invoked when it goes out of scope.

Personally this is why i prefer a `defer` statement so we could do `defer () => lock.unlock()` but that is off topic, but more clear what happens.

**Comment by Josh-Cena** [comment]
https://github.com/tc39/proposal-discard-binding exists. It won't come in a few years but it means this is a legitimate use case. I do think that either ESLint should fix this eventually (not us), or we should just tell everyone to use a special ignore pattern, or use disable comments.

**Comment by JoshuaKGoldberg** [comment]
👍 that this isn't an "us" thing. Thanks for filing though!

...and also +1 on the discard binding proposal, that'd be great.
